### DISCUSSION TITLE
Is there a way to modify the keepAliveTimeout without using a custom server?

### DISCUSSION THREAD

**Discussion by jordie23** [vercel/next.js#19618] [Help]
NodeJS defaults `keepAliveTimeout` to 5 seconds. When NextJS is behind an AWS ELB/ALB, it can cause occasional 502 errors as the ELB's timeout is longer than Next's, which it should be the other way around. Also need to be able to configure `headersTimeout` for this circumstance. 

More info can be found here: https://shuheikagawa.com/blog/2019/04/25/keep-alive-timeout/

**Comment by jordie23** [comment]
I'm guessing not :( Looks like it needs to be set around here and doesn't seem like there's a way to set it: https://github.com/vercel/next.js/blob/764edc51b2ba9ae18046804d3ef9adbe0ea703a0/packages/next/server/lib/start-server.ts#L17

Maybe I'll make a feature request. Would rather not lose features to set up a custom server just to get nextjs working properly on AWS.

**Reply by tabuchid** [reply]
Added a feature request. https://github.com/vercel/next.js/issues/28633

**Comment by jordie23** [comment]
Heya the only solution I was able to find was to modify the IdleTimeout on the AWS ELB to be 4 seconds, i.e 1 second less than node's default 5 seconds.

Which isn't optimal I don't think.

**Comment by 204Constie** [comment]
As the question was not answered and the issue persists i wanted to revive it. Has anyone been able to change the default keep-alive timeout for the server ?

**Comment by anuragkalia** [comment]
I am also facing the same problem. I can only tweak port and hostname for the server and nothing else. This is the only issue which may potentially force us to move to a custom server, which we really do not want to maintain.

**Comment by jordie23** [comment]
Has anyone been able to find a solution to this?

**Comment by rawpixel-vincent** [comment]
We're using patch-package to fix this issue, please make `server.keepAliveTimeout` a next.js settings 🙏🏻 
edit: steps outlined here https://github.com/vercel/next.js/discussions/19618#discussioncomment-1703461

**Reply by rawpixel-vincent** [reply]
https://github.com/vercel/next.js/discussions/22211

**Reply by Especially** [reply]
@rawpixel-vincent Mind outlining the steps involved?

**Reply by rawpixel-vincent** [reply]
1) in your project, edit the file `node_modules/next/dist/server/lib/start-server.js` line 31 add: `srv.keepAliveTimeout = 61000;` and `srv.headersTimeout = 62000;` (mind that the timeout must be higher that what you configure as idle_timeout for your alb, default is 60 seconds)
2) add to your dev dependencies `patch-package` e.g. `yarn add -D patch-package`
3) run `npx patch-package next` to create the patch
4) add to your postinstall scripts in the package.json:
```
 "scripts": {
+  "postinstall": "patch-package"
 }
``` 
5) commit the file created in `/patches`
6) test by deleting your node_modules folder and run `yarn install` the patch should be applied
n.b. you need to regenerate the patch every time you upgrade next, but patch-package will show a warning with the steps to do it.

here's an example of the patch generated for next 11.1.2
```
diff --git a/node_modules/next/dist/server/lib/start-server.js b/node_modules/next/dist/server/lib/start-server.js
index 8edf53d..0ab11a4 100644
--- a/node_modules/next/dist/server/lib/start-server.js
+++ b/node_modules/next/dist/server/lib/start-server.js
@@ -22,6 +22,15 @@ async function start(serverOptions, port, hostname) {
         srv.on('listening', ()=>resolve()
         );
         srv.listen(port, hostname);
+
+        // https://github.com/vercel/next.js/discussions/19618
+        // Ensure all inactive connections are terminated by the ALB,
+        // by setting this a few seconds higher than the ALB idle timeout
+        // https://docs.aws.amazon.com/elasticloadbalancing/latest/application/application-load-balancers.html#connection-idle-timeout
+        // Hint: the alb is pre-opening connections behind the scene
+        srv.keepAliveTimeout = 61000;
+        srv.headersTimeout = 62000;
+
     });
     // It's up to caller to run `app.prepare()`, so it can notify that the server
     // is listening before starting any intensive operations.
```

**Reply by Especially** [reply]
That’s excellent cheers for this! For those who Dockerize their NextJS projects and need this to run for your CI, make sure to use

```
ADD [“patches”, ${WORKDIR}/patches]
```

Before running `npm install`

**Comment by jordie23** [comment]
We are doing something similar to patch, just an `awk` on the bundle after build. As patch replaces the whole file (IIRC) it’s more likely to break things than a regex search and insert.

**Reply by rawpixel-vincent** [reply]
thanks for the advice - first look at the codebase, it's not straightforward for a noob to PR a new config option for this - but we're looking into that. On future version I'll look into `awk` or implement a custom server 😢

**Reply by danroestorf** [reply]
@jordie23 do you have a code example for how you implemented this?

**Comment by tabuchid** [comment]
Added a feature request here: https://github.com/vercel/next.js/issues/28633

**Comment by danroestorf** [comment]
I tested out an extremely hacky solution for this that resolved our 502s. I ran a load test before and after adding this to our Docker build, and the results are definitive: no more 502s. 
```
sed -i "/^srv.on('error',reject)/i srv.keepAliveTimeout = 61000; srv.headersTimeout = 62000;" node_modules/next/dist/server/lib/start-server.js
```

**Disclaimer:** You probably want to work this example a bit before using it in production.

**Comment by tabuchid** [comment]
@jordie23  @eretica and anyone else interested. @timneutkens has re-evaluated and is open to a PR that fixes this. https://github.com/vercel/next.js/discussions/28642#discussioncomment-1716954
I'm not sure when I'll have the capacity to take a look so dropping this here if anyone is interested.

**Comment by dardanos** [comment]
I'm running next@12.3.1 and building the standalone project with docker. I took an approach similar to @DanRoestorf. Sharing it in case someone needs a hand:

```dockerfile
...
RUN npm run build
RUN echo "server.keepAliveTimeout = 65000; server.headersTimeout = 66000;" >> ./.next/standalone/server.js
...
```

**Comment by ziggy6792** [comment]
Hey guys.

Using next 12 it seem like there is a `keepAliveTimeout` option to configure this.
Am I right?
Does this mean I don't have to do this hack?
Does anyone know how I can use this option


`node_modules/next/dist/server/lib/start-server.js`
```
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.startServer = startServer;
var _log = require("../../build/output/log");
var _http = _interopRequireDefault(require("http"));
var _next = _interopRequireDefault(require("../next"));
function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function startServer(opts) {
    let requestHandler;
    const server = _http.default.createServer((req, res)=>{
        return requestHandler(req, res);
    });
    if (opts.keepAliveTimeout) {
        server.keepAliveTimeout = opts.keepAliveTimeout;
    }
    return new Promise((resolve, reject)=>{
        let port = opts.port;
        let retryCount = 0;
        server.on("error", (err)=>{
            if (port && opts.allowRetry && err.code === "EADDRINUSE" && retryCount < 10) {
                (0, _log).warn(`Port ${port} is in use, trying ${port + 1} instead.`);
                port += 1;
                retryCount += 1;
                server.listen(port, opts.hostname);
            } else {
                reject(err);
            }
        });
        let upgradeHandler;
        if (!opts.dev) {
            server.on("upgrade", (req, socket, upgrade)=>{
                upgradeHandler(req, socket, upgrade);
            });
        }
        server.on("listening", ()=>{
            const addr = server.address();
            const hostname = !opts.hostname || opts.hostname === "0.0.0.0" ? "localhost" : opts.hostname;
            const app = (0, _next).default({
                ...opts,
                hostname,
                customServer: false,
                httpServer: server,
                port: addr && typeof addr === "object" ? addr.port : port
            });
            requestHandler = app.getRequestHandler();
            upgradeHandler = app.getUpgradeHandler();
            resolve(app);
        });
        server.listen(port, opts.hostname);
    });
}

//# sourceMappingURL=start-server.js.map
```

**Reply by ziggy6792** [reply]
Oh I think I found it here https://github.com/vercel/next.js/pull/35827/files

so just need `next start --keepAliveTimeout 61000`

Can anyone confirm if this will do?

**Reply by sladg** [reply]
This seems to be outdated on latest version.

**Reply by bmoyroud** [reply]
@sladg does setting `process.env.KEEP_ALIVE_TIMEOUT` to 61000 work for you?

Running `npm run build` generates the below in `.next/standalone/server.js`
<img width="659" alt="image" src="https://github.com/vercel/next.js/assets/9058670/d86bbbc7-b15e-471e-a975-fb8258bad758">

Tested on Next v14.0.3.
